### PR TITLE
feat(compass-sidebar): Add loading states for collection in sidebar COMPASS-5211

### PR DESCRIPTION
**Pull request by gribnoysup**
## Description

This PR extracts sidebar navigation into its own component that is fully detached from the application state and rewritten with keyboard support and using leafygreen components and styles where possible. Here's how it looks:

|Mouse Navigation|Keyboard Navigation|Async Collections|
|---|---|---|
|sidebar1|sidebar2|sidebar-async|

This PR doesn't flip the switch on the global overlay removal yet as we need the databases-collections change to also be ready before this happens, so if you want to try this new sidenav locally with async you need to set the `COMPASS_NO_GLOBAL_OVERLAY` env var to `true`

[Direct link to the navigation tree component because github collapsed it](https://github.com/mongodb-js/compass/blob/9f4d84a2d3d2839c74c86cb430d4657505c6cd89/packages/compass-databases-navigation/src/databases-navigation-tree.tsx) (tell me if you think I should split it into more parts, so far I extracted only things that we might use outside of the component)

### Checklist

- [x] Add unit tests for new components
- [x] Check the checks
- [x] Update e2e tests for the new sidebar
- [ ] Animation for the placeholder (will do in a separate PR so that I can have a shared style between sidebar and databases-collections)
- [x] All namespace actions should work as before
  - [x] select-database
  - [x] drop-database
  - [x] select-collection
  - [x] create-collection
  - [x] drop-collection
  - [x] open-in-new-tab
  - [x] duplicate-view
  - [x] modify-view

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [x] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-components/src/components/databases-navigation-tree.tsx
Comment (gribnoysup) [inline packages/compass-components/src/components/databases-navigation-tree.tsx:199]: This is changed from previous behavior where we would just show some items as disabled, but my thinking here is that if it's just one item left in the context menu, why not just show it right away? I do need to check this with Claudia still

File: packages/compass-components/src/components/databases-navigation-tree.tsx
Comment (gribnoysup) [inline packages/compass-components/src/components/databases-navigation-tree.tsx:387]: As we are now on a pretty modern version of Chromium I though that we might try using [content-visibility](https://developer.mozilla.org/en-US/docs/Web/CSS/content-visibility) instead of the react-virtualized. I didn't notice any significant performance degradation, even the opposite as we are doing much less relayouts now and don't need to watch for the container size changes, but please try it out locally and tell me if you see any weird behaviors

(I also probably need to add this to the collection list explicitly as the user might have small amount of databases with enormous amounts of collections)

File: packages/compass-components/src/components/databases-navigation-tree.spec.tsx
Comment (mcasimir) [inline packages/compass-components/src/components/databases-navigation-tree.spec.tsx:1]: Having this component here implies a broader architectural decision, would be great if compass/components could only contain domain-free components and raw building blocks, that's how we are approaching this so far at least.

Breaking this pattern has a good potential to create confusion and we won't have anymore a clear cut as for when to put something in components and when not. For example by the same token the connection form could have ended up here. 

Is there a reason (other than ts) why this should not be in the sidebar?

Comment (gribnoysup) [reply packages/compass-components/src/components/databases-navigation-tree.spec.tsx:1]: I was assuming the only goal with this package is not to have any direct connection to compass internals (which some hadron components have for example), thanks for clarifying that! The main reason was indeed that converting the whole sidebar plugin codebase to the new ts, code formatting, linting, and testing setup would be too much work upfront and as this side navigation is just a react component, putting it here made sense to me. I'm happy to move all this to the separate new monorepo package or work on converting the plugin to our new tooling, whatever makes more sense in your opinion.

I'm also totally okay with following the "only contain domain-free components and raw building blocks", but not really sure though why we are doing it. From my POV if keeping plugins (and tbh most of the other packages in the monorepo) in separate packages was more or less a decision based on how they were published before, consumed by Compass, and the fact that they have slightly different build/test setups, for the new components I don't really see a reason to create multiple packages like that, so I wonder what might be the motivation that I'm missing.

Comment (mcasimir) [reply packages/compass-components/src/components/databases-navigation-tree.spec.tsx:1]: Yeah I totally get your point, if is not a big deal for you I'd stick to the pattern where compass-components just domain-free for now. Practically speaking honestly is fine to do this in a separate PR.

I don't mean to be too annoying here, and we can change this rule, but this is my rationale: despite of how good this convention is, and how fit is the current "plugin-based" design with the monorepo, for me we should stick with it until we consciously decide to change, and if and when we do, we should change things everywhere and consistently.

To me is really important that we don't take this decision lightly, and we don't make it happen by accident or as a byproduct of something else (ie `ts`) cause essentially it would create a situation where 2 different "frontends" would exist, one in the plugins and one in `compass-components`, and it since it won't be fun to undo it ..  if that's the direction we want to take we should also make a commitment and a plan to move to only one, at least not leaving it to the case.

> I'm totally okay with following the "only contain domain-free components and raw building blocks", but not really sure though why we are doing it. From my POV if keeping plugins (and tbh most of the other packages in the monorepo) in separate packages was more or less a decision based on how they were published before, consumed by Compass, and the fact that they have slightly different build/test setups, for the new components I don't really see a reason to create multiple packages like that, so I wonder what might be the motivation that I'm missing.

I think here there are 2 design decisions involved though:
- The first decision is for `compass-components`: this one is basically a central place for reusable UI components, is a way to change in one place how the application looks, kinda a re-export from LG + the components we miss from LG, but that's the level and purpose, just a way to factor together duplicate styles.

- The second one is about the plugins: i don't think the separation they create is so terrible, but i get that we don't need to have them in 100 packages. Also we have many that do not make so much sense. The general fact that we have things with significantly different state / doing different actions and with an independent set of functionalities separated makes sense to me, but yeah those could be just different folders in a package rather then different packages and i'd be ok with that.

Personally i think those should be 2 different decisions, for example if we remove the packages then we could make everything just a folder inside the same package, but `compass-components` could still be a subfolder, we would import `<Button>` and `colors.gray1` from there and `sidebar` could be another one.

Comment (gribnoysup) [reply packages/compass-components/src/components/databases-navigation-tree.spec.tsx:1]: > if is not a big deal for you I'd stick to the pattern where compass-components just domain-free for now

Yeah, absolutely, will move it out, thanks for sharing your thoughts on this.

Comment (gribnoysup) [reply packages/compass-components/src/components/databases-navigation-tree.spec.tsx]: Done in 22c421e, but tbh I'm not sure if I should've moved more or less stuff, in theory useTree hooks are pretty generic, but then also would we really have another navigation tree view in the app? Idk, anyway, tell me what you think and if this matches what you were expecting

File: packages/compass-components/src/components/databases-navigation-tree.tsx
Comment (gribnoysup) [reply packages/compass-components/src/components/databases-navigation-tree.tsx]: Checked with Claudia, keeping this behavior

File: packages/compass-databases-navigation/src/databases-navigation-tree.tsx
Comment (Anemy) [inline packages/compass-databases-navigation/src/databases-navigation-tree.tsx:400]: Currently in Compass the database and collections tree items have a `cursor`: `pointer` on hover, do we want to keep that functionality? I have no preference.

**Review by Anemy — COMMENTED** [review]
This is real cool. Sick cleanup and adding of accessibility. I really like the nav tree component and hooks, solid tests too.
Added a couple comments and noticed a few spots we can remove some old code.

I ran into two things playing around with this branch on my computer:
One was scrolling with a lot of collections in the sidebar was making the scroll jump unexpectedly. This can be seen with the `test` db in the compass-data-sets deployment.

https://user-images.githubusercontent.com/1791149/142952944-ab82baa4-f9e3-49c1-9f3c-c13aa705cae1.mp4


The other issue I noticed was when I searched for dbs/collections with the filter they weren't loading. Is that work planned elsewhere?

Do you have a database with a lot of collections to test the `content-visibility`? I think the main reason why we used `react-virtualized` in the collection/db list was for cases where the user has hundreds or thousands of dbs/collections. Probably would be good to test that manually before merging.

File: packages/collection-model/lib/model.js
Comment (gribnoysup) [reply packages/collection-model/lib/model.js:204]: This is just keeping the compat with what we are using right now as a `source` value when new collection tab is created, changing this will probably touch most of the plugins in compass (at least all collection tab related ones) and I would want to avoid that

File: packages/compass-components/src/components/content-with-fallback.tsx
Comment (gribnoysup) [reply packages/compass-components/src/components/content-with-fallback.tsx:1]: I have types for the components, but this rule still errors out due to propTypes missing, I wasn't sure if I should make a call and disable them globally, but I don't see why I would add them here and what value they bring for typescript components

File: packages/compass-components/src/hooks/use-navigation-tree.ts
Comment (gribnoysup) [reply packages/compass-components/src/hooks/use-navigation-tree.ts:161]: Good catch! Missed this part of the spec, will fix.

Yeah, it's basically catching all the unicode symbols that have LETTER property. This `\p` is called [unicode property escape](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Unicode_Property_Escapes) and it allows you to match by unicode properties. There is a ton of cases where it can simplify targeting your regex, for example you can catch all emojis with `{Emoji_Presentation}` group

File: packages/compass-databases-navigation/src/databases-navigation-tree.tsx
Comment (gribnoysup) [reply packages/compass-databases-navigation/src/databases-navigation-tree.tsx:486]: So the reason I kept them in one file is that they are not used outside and they do have overlapping styles that are shared, I was trying to avoid having e.g., a five line file for `actionsContainer` styles that are shared between databases and collections, for me personally it creates a ton of load if I need to jump to a file just to see these. As I mentioned in the PR description I don't mind splitting them though if this feels like too much for one

File: packages/compass-collection/src/modules/tabs.js
Comment (gribnoysup) [reply packages/compass-collection/src/modules/tabs.js:545]: This `selectOrCreateTab` method uses both `createNewTab` and `replaceTabContent`, it also the only method that is always called when tabs are switching, this is why I put it here, I can either move the call out to those methods separately, or keep it here, what do you prefer?

File: packages/compass-databases-navigation/src/databases-navigation-tree.tsx
Comment (gribnoysup) [reply packages/compass-databases-navigation/src/databases-navigation-tree.tsx:176]: Not sure I follow. I might be talking about something different to what you meant, but this `isReadOnly` property here refers to instance or compass being readonly (due to topology or read only edition), not to the view itself being readonly and you probably shouldn't be able to modify the view in that case, right?

File: packages/compass-components/src/hooks/use-navigation-tree.ts
Comment (gribnoysup) [reply packages/compass-components/src/hooks/use-navigation-tree.ts:77]: IIRC `activeElement` will always be set to something and "no focus" is when it is equal to body. We can definitely short circuit by either checking if it's `body` first or even checking if our root node ref [contains](https://developer.mozilla.org/en-US/docs/Web/API/Node/contains) the active element. I'm wondering though why do you think we need a short circuit here?

**Comment by gribnoysup** [comment]
> Do you have a database with a lot of collections to test the content-visibility? I think the main reason why we used react-virtualized in the collection/db list was for cases where the user has hundreds or thousands of dbs/collections. Probably would be good to test that manually before merging.

So `content-visibility` is a property that does same thing that react-virtualized does just natively by the browser. Yes, as I mentioned I tested it with both a database that has 10k collections and a server that has ~1k databases (my local setup couldn't handle creating more than that), the performance looked similar to me (the rendering of the sidebar is mostly fine, but generally it's bad both in this branch and in current compass, for a different reason though, see COMPASS-5316), but I am not getting the scroll jitter so I'll change it back to react-virtualized (or whatever it was deprecated in favor of, seems like it's react-window now?)

**Comment by gribnoysup** [comment]
> The other issue I noticed was when I searched for dbs/collections with the filter they weren't loading. Is that work planned elsewhere?

Is this with or without global overlay? If without then it's kinda expected for now, but I might just fix it in this branch so it's not confusing

File: packages/compass-components/src/hooks/use-navigation-tree.ts
Comment (Anemy) [reply packages/compass-components/src/hooks/use-navigation-tree.ts:77]: All good

File: packages/compass-databases-navigation/src/databases-navigation-tree.tsx
Comment (Anemy) [reply packages/compass-databases-navigation/src/databases-navigation-tree.tsx:176]: I think re-discussions today the only thing maybe we can address here is that `datalake` gets marked as readonly, however you can create views in datalake (but can't modify them apparently). 
I think no changes needed here 👍 
Definitely worth chatting about sometime how we can make what read only restriction is applied more transparent in Compass.

File: packages/compass-databases-navigation/src/databases-navigation-tree.tsx
Comment (Anemy) [reply packages/compass-databases-navigation/src/databases-navigation-tree.tsx:486]: That's fair, I think a few of the styles, like the focus ring, could be something we share with the connections sidebar. Fine as it is, but if we find we're adding more logic here I think it's definitely worth breaking into more files.

**Comment by Anemy** [comment]
>>The other issue I noticed was when I searched for dbs/collections with the filter they weren't loading. Is that work planned elsewhere?

> Is this with or without global overlay? If without then it's kinda expected for now, but I might just fix it in this branch so it's not confusing

Without. Cool, no worries, just making sure we're aware of it.

**Comment by gribnoysup** [comment]
@Anemy this turned out to be an almost full rewrite to make it accessible with react-virtualized so even though I'm still cleaning up stuff and writing tests, maybe you already want to take a look if you have a moment?

File: packages/compass-components/src/hooks/use-navigation-tree.spec.tsx
Comment (gribnoysup) [inline packages/compass-components/src/hooks/use-navigation-tree.spec.tsx:1]: This needs to be removed and replaced with the one that is currently in the databases-navigation package directly

File: packages/compass-databases-navigation/src/databases-navigation-tree.tsx
Comment (Anemy) [inline packages/compass-databases-navigation/src/databases-navigation-tree.tsx:367]: In the interest of making Compass more web compatible, would it make sense to have these as optional props or have them polluting the global namespace a bit for tests instead of process env variables? Fine how it is for Compass at the moment, just a thought.

**Review by Anemy — APPROVED** [review]
Good stuff. Tested it out a bit and works so smooth. Feels solid to the touch. Nice work on the accessibility improvements. Good tests and nice comments on some of the trickier things with relevant links 👍 
Couple questions/ideas/suggestions, but no blockers. Looks good to go once tests are passing!

File: packages/compass-databases-navigation/src/databases-navigation-tree.tsx
Comment (gribnoysup) [reply packages/compass-databases-navigation/src/databases-navigation-tree.tsx:367]: Yeah, totally, I can move them to props

**Comment by gribnoysup** [comment]
> I'm thinking it probably makes sense to move it to them into both createNewTab and replaceTabContent since we can open up a collection view from 

Oh, I totally missed that spot, will fix!

File: packages/compass-app-stores/src/stores/instance-store.js
Comment (gribnoysup) [reply packages/compass-app-stores/src/stores/instance-store.js:65]: > Small thing, but how do you feel about having these states as constants/an enum we can refer to?

I would like to, but I legit don't know where to put them and I don't want to move the models in one package at least in the scope of this work if that's okay (I might do this as a separate PR or something, there is enough reasons for that already, sounds alright?)

> Could we potentially be redoing the same query many times?

All inflight requests for the model data are currently debounced (for all the instance related models, but I'll link only to one code example), so if multiple parts of the application are calling `instance.fetch`, they will be awaiting on the same request promise:

https://github.com/mongodb-js/compass/blob/a49978800cfefbd05b71e93b0671adf67a960b24/packages/instance-model/lib/model.js#L6-L35

This particular check here doesn't mean that we start another fetch while the current one is in progress, it just guarantees we wait for the in-progress one to finish before starting to fetch collection info.

Giving it another look I am thinking that I'll probably move this to the models, similar how the whole `instance.refresh` flow is now part of the model so that the stores are not doing much except for just calling model methods directly

Comment (gribnoysup) [reply packages/compass-app-stores/src/stores/instance-store.js]: Done in 5548eec04ba6dc207dc109c2a8e85fb02b1c9e91

File: packages/compass-components/src/hooks/use-focus-hover.ts
Comment (gribnoysup) [reply packages/compass-components/src/hooks/use-focus-hover.ts]: Removed in ba0aafa6ed27effa077a5888bd889f64282d30f5

File: packages/compass-databases-navigation/src/databases-navigation-tree.tsx
Comment (gribnoysup) [reply packages/compass-databases-navigation/src/databases-navigation-tree.tsx]: Done in a2405d7

**Pull request merged**
